keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,740 @@
+More information about usage of these API types is provided inline with user stories.
 
-```yaml
-apiVersion: cosi.io/v1alpha1
-kind: BucketAccessClass
-metadata: 
-  name:
-provisioner: [1]
-policyActionsConfigMap: [2]
-  name:
-  namespace:
-parameters: [3]
-```
+## Bucket Creation
 
-1. `provisioner`: (optional) the name of the driver that `BucketAccess` instances should be managed by. If empty then there is no driver/sidecar, thus the admin had to create the `Bucket` and `BucketAccess` instances.
-1. `policyActionsConfigMap`: (required) a reference to a ConfigMap that contains a set of provisioner/platform-defined policy actions for bucket access. It is seen as typical that this config map's namespace is the same as for the provisioner. In othe words, a namespace that has locked down RBAC rules or prevent modification of this config map.
-1. `parameters`: (Optional)  A map of string:string key values.  Allows admins to control user and access provisioning by setting provisioner key-values. Optional reserved keys cosi.io/configMap and cosi.io/secrets are used to reference user created resources with provider specific access policies.
----
+The following stakeholders are involved in the lifecycle of bucket creation:
 
-### App Pod
-The application pod utilizes CSI's inline ephemeral volume support to provide the endpoint and secret credentials in an in-memory volume. This approach also, importantly, prevents the pod from launching before the bucket has been provisioned since the kubelet waits to start the pod until it has received the cosi-node-adapter's _NodeStageVolume_ response.
+ - Users  - request buckets to be created for their workload
+ - Admins - setup COSI drivers in the cluster, and specify their configurations
 
-Here is a sample pod manifest:
+Here are the steps for creating a Bucket:
 
-```yaml
-apiVersion: v1
-kind: Pod
-metadata:
-  name: my-app-pod
-  namespace: dev-ns
-spec:
-  serviceAccountName: [1]
-  containers:
-    - name: my-app
-      image: ...
-      volumeMounts:
-        - mountPath: /cosi [2]
-          name: cosi-vol
-  volumes:
-    - name: cosi-vol
-      csi: [3]
-        driver: cosi.sigs.k8s.io [4]
-        volumeAttributes: [5]
-          bucketAccessRequestName: <my-bar-name>
+  ![COSI Bucket Creation](images/create-bucket.jpg)
+
+
+###### 1. Admin creates BucketClass, User requests Bucket to be created
+
+The BucketClass represents a set of common properties shared by multiple buckets. It is used to specify the driver for creating Buckets, and also for configuring driver-specific parameters. More information about BucketClass is [here](#bucketclass)
+
+The BucketRequest is a request to create a Bucket. This resource is designed to specify properties that the application requires, such as the data protocol that this bucket should satisfy (i.e. s3, gcs or azure). More information about BucketRequest is [here](#bucketrequest)
```

Comment:
> Is this information really needed by the user? Could it be part of the provisioner instead? A use case would be if a provisioner can support many protocols but this could also be expressed with two different provisioners each one with its own protocol.

This field is mainly needed to ensure that the data protocol supported by the dynamically created bucket will be the same as what the user application expects. This field has to be specified by the user. This field will also be crucial for cross cluster interoperability.

Reply:
Then the user will not be able to specify what the application needs. The protocol is a opaque field and COSI does no validation of the protocol. The driver is free to do validation if needed. In other Kubernetes resources, the class seems to encode Kubernetes specific information like RetentionPolicy. Resources relevant to the application and not relevant to Kubernetes, say FSType in StorageClasses, seems to be largely unused AFAICT

Reply:
I'n not following - the app expressed what it needs in it's selection of a class.  Or are you planning to auto-select a class based on protocol?  That seems a little specious to me, but I admit I am not an expert here.  I didn't see anything in here to that effect.

IOW if I choose "class gold" which is provider=GCS and I specify protocol=S3, what does that mean?  S3 is not a great example because GCS supports a flavor of S3 API, but you get my point.

Reply:
No, I’m saying classes won’t have anything OSP or protocol specific fields. It’ll only have provider agnostic information - and can be used across providers.

Reply:
Now I am confused - the KEP says "It is used to specify the driver for creating Buckets, and also for configuring driver-specific parameters" and has a `provisioner` field.

This is what classes are for - to swallow implementation details.